### PR TITLE
fix #15215

### LINKED ISSUE #15215: net.connect with timeout parameter broken for SSL sockets
Reported in #15206. This issue is to track it and to have a point of reference for the docs.

### PR DESCRIPTION
**Pull request by rockcavera**
fix [15215](https://github.com/nim-lang/Nim/issues/15215)

In my tests it worked perfectly copying the handshake scheme from the other `net.connect`.

This PR is the same as this [here](https://github.com/nim-lang/Nim/pull/16136). It just so happened that I deleted the old fork and could no longer edit it.

### CODE REVIEW

File: tests/stdlib/tnetconnect.nim [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# Issue 15215 - https://github.com/nim-lang/Nim/issues/15215
+proc test() =
+  var
+    ctx = newContext()
+    socket = newSocket()
+
+  wrapSocket(ctx, socket)
+
+  connect(socket, "www.nim-lang.org", Port(443), 5000)
+
```
Comment (ringabout) [inline tests/stdlib/tnetconnect.nim:17]: needs to send something like the coded posted on Nim forum

```nim
import net

var ctx = newContext()
var next_socket = newSocket()
wrapSocket(ctx, next_socket)

next_socket.connect("www.google.com", Port(443), 1000)

next_socket.send("GET /")
```

https://forum.nim-lang.org/t/6686

**Comment by rockcavera** [comment]
How do I guide the test that needs to compile with `-d:ssl`?

**Comment by ringabout** [comment]
use a `tests/stdlib/tnetconnect.nim.cfg` or `tests/stdlib/tnetconnect.nims` just like what `tests/stdlib/tssl.nims` does?

File: tests/stdlib/tnetconnect.nim [hunk 2]
```diff
@@ -0,0 +1,22 @@
+discard """
+  cmd: "nim c -d:ssl $file"
```
Comment (ringabout) [inline tests/stdlib/tnetconnect.nim]:
```suggestion
  cmd: "nim c -r -d:ssl $file"
```

File: lib/pure/net.nim [hunk 3]
```diff
@@ -2003,7 +1999,18 @@ proc connect*(socket: Socket, address: string, port = Port(0),
...
+        # RFC3546 for SNI specifies that IP addresses are not allowed.
+        if not isIpAddress(address):
+          # Discard result in case OpenSSL version doesn't support SNI, or we're
+          # not using TLSv1+
+          discard SSL_set_tlsext_host_name(socket.sslHandle, address)
+
+        ErrClearError()
+        let ret = SSL_connect(socket.sslHandle)
+        socketError(socket, ret)
+        when not defined(nimDisableCertificateValidation) and not defined(windows):
```
Comment (Araq) [inline lib/pure/net.nim]:
```suggestion
        when not defined(nimDisableCertificateValidation):
```

**Pull request merged**